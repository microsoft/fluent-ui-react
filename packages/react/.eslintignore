src/themes/teams/components/Icon/svg/ProcessedIcons
coverage/
dist/
lib/
node_modules/
