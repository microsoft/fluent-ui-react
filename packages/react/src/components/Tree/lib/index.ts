import * as _ from 'lodash'
import { TreeItemProps } from '../TreeItem'
import { ShorthandValue } from '../../../types'

const hasSubtree = (item: TreeItemProps | ShorthandValue<TreeItemProps>): boolean => {
  return !_.isNil(item['items']) && item['items'].length > 0
}

const removeItemAtIndex = (items: any[], itemIndex: number): any[] => {
  return [...items.slice(0, itemIndex), ...items.slice(itemIndex + 1)]
}

export { hasSubtree, removeItemAtIndex }
