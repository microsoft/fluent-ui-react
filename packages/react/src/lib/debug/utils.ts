export const containsSubstring = (arg: any, substring: string) => {
  return typeof arg === 'string' && arg.indexOf(substring) >= 0
}

export const getLastOf = function<Item>(array: Item[], fallback: Item) {
  return array.length > 0 ? array[array.length - 1] : fallback
}

export const isNotNullOrEmpty = object => {
  if (object == null) {
    return false
  }

  return Object.keys(object).length > 0
}

export const isNotEmptyObjectsArray = array => {
  if (array == null) {
    return false
  }

  return array.some(isNotNullOrEmpty)
}

export const deepPickBy = (
  object: any,
  predicate: (propName: string, propValue: any) => boolean,
  seenObjects = [],
) => {
  if (!object) {
    return {}
  }

  if (seenObjects.some(seenObject => seenObject === object)) {
    throw new Error(`Circular dependency detected.`)
  }

  return Object.keys(object).reduce((acc, currentPropName) => {
    if (predicate(currentPropName, object[currentPropName])) {
      return { ...acc, [currentPropName]: object[currentPropName] }
    }

    if (typeof object[currentPropName] === 'object') {
      const deepPickResult = deepPickBy(object[currentPropName], predicate, [
        ...seenObjects,
        object,
      ])

      if (isNotNullOrEmpty(deepPickResult)) {
        return { ...acc, [currentPropName]: deepPickResult }
      }
    }

    return acc
  }, {})
}

export const deepPick = (object, propName) =>
  deepPickBy(object, currentPropName => propName === currentPropName)
