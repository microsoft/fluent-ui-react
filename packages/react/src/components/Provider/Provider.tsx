import { IStyle } from 'fela'
import { render } from 'fela-dom'
import * as _ from 'lodash'
import * as PropTypes from 'prop-types'
import * as React from 'react'
import { RendererProvider, ThemeProvider } from 'react-fela'

import { felaRenderer as felaLtrRenderer, isBrowser, mergeThemes } from '../../lib'
import {
  ThemePrepared,
  ThemeInput,
  StaticStyleObject,
  StaticStyle,
  StaticStyleFunction,
  FontFace,
} from '../../themes/types'
import ProviderConsumer from './ProviderConsumer'
import { mergeSiteVariables } from '../../lib/mergeThemes'

export interface ProviderProps {
  theme: ThemeInput
  children: React.ReactNode
}

/**
 * The Provider passes the CSS in JS renderer and theme to your components.
 */
class Provider extends React.Component<ProviderProps> {
  staticStylesRendered: boolean = false

  static propTypes = {
    theme: PropTypes.shape({
      siteVariables: PropTypes.object,
      componentVariables: PropTypes.object,
      componentStyles: PropTypes.object,
      rtl: PropTypes.bool,
      fontFaces: PropTypes.arrayOf(
        PropTypes.shape({
          name: PropTypes.string,
          paths: PropTypes.arrayOf(PropTypes.string),
          style: PropTypes.shape({
            fontStretch: PropTypes.string,
            fontStyle: PropTypes.string,
            fontVariant: PropTypes.string,
            fontWeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
            localAlias: PropTypes.string,
            unicodeRange: PropTypes.string,
          }),
        }),
      ),
      staticStyles: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.func]),
      ),
      animations: PropTypes.object,
    }),
    children: PropTypes.element.isRequired,
  }

  static Consumer = ProviderConsumer

  renderStaticStyles = (mergedTheme: ThemePrepared) => {
    // RTL WARNING
    // This function sets static styles which are global and renderer agnostic
    // With current implementation, static styles cannot differ between LTR and RTL
    // @see http://fela.js.org/docs/advanced/StaticStyle.html for details

    const { siteVariables } = mergedTheme
    const { staticStyles } = this.props.theme

    if (!staticStyles) return

    const renderObject = (object: StaticStyleObject) => {
      _.forEach(object, (style, selector) => {
        felaLtrRenderer.renderStatic(style as IStyle, selector)
      })
    }

    staticStyles.forEach((staticStyle: StaticStyle) => {
      if (typeof staticStyle === 'string') {
        felaLtrRenderer.renderStatic(staticStyle)
      } else if (_.isPlainObject(staticStyle)) {
        renderObject(staticStyle as StaticStyleObject)
      } else if (_.isFunction(staticStyle)) {
        const preparedSiteVariables = mergeSiteVariables(siteVariables)
        renderObject((staticStyle as StaticStyleFunction)(preparedSiteVariables))
      } else {
        throw new Error(
          `staticStyles array must contain CSS strings, style objects, or style functions, got: ${typeof staticStyle}`,
        )
      }
    })
  }

  renderFontFaces = () => {
    // RTL WARNING
    // This function sets static styles which are global and renderer agnostic
    // With current implementation, static styles cannot differ between LTR and RTL
    // @see http://fela.js.org/docs/advanced/StaticStyle.html for details

    const { fontFaces } = this.props.theme

    if (!fontFaces) return

    const renderFontObject = (font: FontFace) => {
      if (!_.isPlainObject(font)) {
        throw new Error(`fontFaces must be objects, got: ${typeof font}`)
      }
      felaLtrRenderer.renderFont(font.name, font.paths, font.style)
    }

    fontFaces.forEach((font: FontFace) => {
      renderFontObject(font)
    })
  }

  componentDidMount() {
    this.renderFontFaces()
  }

  render() {
    const { theme, children } = this.props

    // rehydration disabled to avoid leaking styles between renderers
    // https://github.com/rofrischmann/fela/blob/master/docs/api/fela-dom/rehydrate.md
    return (
      <ProviderConsumer
        render={(incomingTheme: ThemePrepared) => {
          const outgoingTheme: ThemePrepared = mergeThemes(incomingTheme, theme)

          // Heads up!
          // We should call render() to ensure that a subscription for DOM updates was created
          // https://github.com/stardust-ui/react/issues/581
          if (isBrowser()) render(outgoingTheme.renderer)
          this.renderStaticStylesOnce(outgoingTheme)

          return (
            <RendererProvider renderer={outgoingTheme.renderer} {...{ rehydrate: false }}>
              <ThemeProvider theme={outgoingTheme}>{children}</ThemeProvider>
            </RendererProvider>
          )
        }}
      />
    )
  }

  renderStaticStylesOnce = (mergedTheme: ThemePrepared) => {
    const { staticStyles } = this.props.theme
    if (!this.staticStylesRendered && staticStyles) {
      this.renderStaticStyles(mergedTheme)
      this.staticStylesRendered = true
    }
  }
}

export default Provider
