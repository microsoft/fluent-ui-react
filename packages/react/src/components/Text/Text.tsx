import * as PropTypes from 'prop-types'
import * as React from 'react'

import {
  childrenExist,
  createShorthandFactory,
  customPropTypes,
  UIComponent,
  UIComponentProps,
  ContentComponentProps,
  ChildrenComponentProps,
  commonPropTypes,
  ColorComponentProps,
  rtlTextContainer,
  SizeValue,
} from '../../lib'
import { Accessibility } from '../../lib/accessibility/types'
import { defaultBehavior } from '../../lib/accessibility'
import { ReactProps } from '../../types'

export interface TextProps
  extends UIComponentProps,
    ContentComponentProps,
    ChildrenComponentProps,
    ColorComponentProps {
  /**
   * Accessibility behavior if overridden by the user.
   * @default defaultBehavior
   */
  accessibility?: Accessibility

  /** At mentions can be formatted to draw users' attention. Mentions for "me" can be formatted to appear differently. */
  atMention?: boolean | 'me'

  /** Set as disabled Text component */
  disabled?: boolean

  /** Set as error Text component */
  error?: boolean

  /** The text can appear more important and draw user's attention */
  important?: boolean

  /** The size for the Text component */
  size?: SizeValue

  /** The weight for the Text component */
  weight?: 'light' | 'semilight' | 'regular' | 'semibold' | 'bold'

  /** Set as success Text component */
  success?: boolean

  /** The text can signify a temporary state */
  temporary?: boolean

  /** Set as timestamp Text component */
  timestamp?: boolean

  /** Truncates text as needed */
  truncated?: boolean
}

/**
 * A Text component formats occurrences of text consistently.
 * @accessibility
 * Text is how people read the content on your website.
 * Ensure that a contrast ratio of at least 4.5:1 exists between text and the background behind the text.
 *
 * To ensure that RTL mode will be properly handled for provided 'content' value, ensure that either:
 * - 'content' is provided as plain string (then dir="auto" attribute will be applied automatically)
 * - for other 'content' value types (i.e. that use elements inside) ensure that dir="auto" attribute is applied for all places in content where necessary
 */
class Text extends UIComponent<ReactProps<TextProps>, any> {
  static create: Function

  static className = 'ui-text'

  static displayName = 'Text'

  static propTypes = {
    ...commonPropTypes.createCommon({ color: true }),
    atMention: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['me'])]),
    disabled: PropTypes.bool,
    error: PropTypes.bool,
    important: PropTypes.bool,
    size: customPropTypes.size,
    weight: PropTypes.oneOf(['light', 'semilight', 'regular', 'semibold', 'bold']),
    success: PropTypes.bool,
    temporary: PropTypes.bool,
    timestamp: PropTypes.bool,
    truncated: PropTypes.bool,
  }

  static defaultProps = {
    accessibility: defaultBehavior,
    as: 'span',
  }

  renderComponent({ accessibility, ElementType, classes, unhandledProps }): React.ReactNode {
    const { children, content } = this.props

    return (
      <ElementType
        className={classes.root}
        {...rtlTextContainer.getAttributes({ forElements: [children, content] })}
        {...accessibility.attributes.root}
        {...unhandledProps}
      >
        {childrenExist(children) ? children : content}
      </ElementType>
    )
  }
}

Text.create = createShorthandFactory(Text, 'content')

export default Text
