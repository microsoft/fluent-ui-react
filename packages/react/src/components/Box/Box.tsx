import * as React from 'react'
import {
  childrenExist,
  createShorthandFactory,
  UIComponentProps,
  ContentComponentProps,
  ChildrenComponentProps,
  commonPropTypes,
  rtlTextContainer,
} from '../../lib'
import createComponent, { CreateComponentReturnType } from '../../lib/createComponent'
import { ReactProps } from '../../../types/utils'

export interface BoxProps
  extends UIComponentProps<BoxProps>,
    ContentComponentProps,
    ChildrenComponentProps {}

/**
 * A Box is a basic component (no default styles)
 */
const Box: CreateComponentReturnType<ReactProps<BoxProps>> = createComponent<BoxProps>({
  displayName: 'Box',

  className: 'ui-box',

  propTypes: {
    ...commonPropTypes.createCommon(),
  },

  render(config, props) {
    const { ElementType, classes, unhandledProps } = config
    const { children, content } = props

    return (
      <ElementType
        {...rtlTextContainer.getAttributes({ forElements: [children, content] })}
        {...unhandledProps}
        className={classes.root}
      >
        {childrenExist(children) ? children : content}
      </ElementType>
    )
  },
})

Box.create = createShorthandFactory(Box)

export default Box
