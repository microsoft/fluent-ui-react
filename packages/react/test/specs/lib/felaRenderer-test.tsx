import * as React from 'react'
import { createSnapshot } from 'jest-react-fela'
import { EmptyThemeProvider } from 'test/utils'
import Box from 'src/components/Box/Box'
import Provider from 'src/components/Provider/Provider'
import Text from 'src/components/Text/Text'
import { felaRenderer } from 'src/lib'

test('basic styles', () => {
  const snapshot = createSnapshot(
    <EmptyThemeProvider>
      <Box styles={{ color: 'red' }} />
    </EmptyThemeProvider>,
    {},
    felaRenderer,
  )
  expect(snapshot).toMatchSnapshot()
})

test('css fallback value', () => {
  const snapshot = createSnapshot(
    <EmptyThemeProvider>
      <Box styles={{ color: ['red', 'blue'] }} />
    </EmptyThemeProvider>,
    {},
    felaRenderer,
  )
  expect(snapshot).toMatchSnapshot()
})

test('keyframe', () => {
  const spinner = {
    keyframe: ({ fromColor, toColor }) => ({
      from: {
        color: fromColor,
      },
      to: {
        color: toColor,
      },
    }),
    keyframeParams: {
      fromColor: 'red',
      toColor: 'blue',
    },
    duration: '5s',
  }

  const snapshot = createSnapshot(
    <Provider theme={{ animations: { spinner } }}>
      <Box animation="spinner" />
    </Provider>,
    {},
    felaRenderer,
  )
  expect(snapshot).toMatchSnapshot()
})

test('array returned by keyframe results in CSS fallback values', () => {
  const steps = ['0%', '100%']

  const spinner = {
    keyframe: () => {
      const obj = {}
      steps.forEach((step: string, index) => {
        ;(obj as any)[step] = { color: ['blue', 'red', 'yellow'] }
      })
      return obj
    },
  }

  const snapshot = createSnapshot(
    <Provider theme={{ animations: { spinner } }}>
      <Box animation="spinner" />
    </Provider>,
    {},
    felaRenderer,
  )
  expect(snapshot).toMatchSnapshot()
})

test('rtl', () => {
  const snapshot = createSnapshot(
    <Provider rtl={true}>
      <Text content="Hello" styles={{ marginLeft: '10px' }} />
    </Provider>,
    {},
    felaRenderer,
  )
  expect(snapshot).toMatchSnapshot()
})
