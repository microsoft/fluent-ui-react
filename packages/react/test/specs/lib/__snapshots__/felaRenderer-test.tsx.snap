// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css fallback value 1`] = `
".a {
  color: red;
}


<div className=ui-box a />;
"
`;

exports[`keyframe 1`] = `
"@font-face {
  font-weight: 400;
  src: url(\\"https://static2.sharepointonline.com/files/fabric/assets/fonts/segoeui-westeuropean/segoeui-regular.woff2\\")
    format(\\"woff2\\");
  font-family: \\"Segoe UI\\";
}
@font-face {
  font-weight: 600;
  src: url(\\"https://static2.sharepointonline.com/files/fabric/assets/fonts/segoeui-westeuropean/segoeui-semibold.woff2\\")
    format(\\"woff2\\");
  font-family: \\"Segoe UI\\";
}
@font-face {
  font-weight: 700;
  src: url(\\"https://static2.sharepointonline.com/files/fabric/assets/fonts/segoeui-westeuropean/segoeui-bold.woff2\\")
    format(\\"woff2\\");
  font-family: \\"Segoe UI\\";
} /*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */
html {
  line-height: 1.15;
  -webkit-text-size-adjust: 100%;
}
body {
  margin: 0;
}
main {
  display: block;
}
h1 {
  font-size: 2em;
  margin: 0.67em 0;
}
hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}
pre {
  font-family: monospace, monospace;
  font-size: 1em;
}
a {
  background-color: transparent;
}
abbr[title] {
  border-bottom: none;
  text-decoration: underline;
  text-decoration: underline dotted;
}
b,
strong {
  font-weight: bolder;
}
code,
kbd,
samp {
  font-family: monospace, monospace;
  font-size: 1em;
}
small {
  font-size: 80%;
}
sub,
sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
sub {
  bottom: -0.25em;
}
sup {
  top: -0.5em;
}
img {
  border-style: none;
}
button,
input,
optgroup,
select,
textarea {
  font-family: inherit;
  font-size: 100%;
  line-height: 1.15;
  margin: 0;
}
button,
input {
  overflow: visible;
}
button,
select {
  text-transform: none;
}
[type=\\"button\\"],
[type=\\"reset\\"],
[type=\\"submit\\"],
button {
  -webkit-appearance: button;
}
[type=\\"button\\"]::-moz-focus-inner,
[type=\\"reset\\"]::-moz-focus-inner,
[type=\\"submit\\"]::-moz-focus-inner,
button::-moz-focus-inner {
  border-style: none;
  padding: 0;
}
[type=\\"button\\"]:-moz-focusring,
[type=\\"reset\\"]:-moz-focusring,
[type=\\"submit\\"]:-moz-focusring,
button:-moz-focusring {
  outline: 1px dotted ButtonText;
}
fieldset {
  padding: 0.35em 0.75em 0.625em;
}
legend {
  box-sizing: border-box;
  color: inherit;
  display: table;
  max-width: 100%;
  padding: 0;
  white-space: normal;
}
progress {
  vertical-align: baseline;
}
textarea {
  overflow: auto;
}
[type=\\"checkbox\\"],
[type=\\"radio\\"] {
  box-sizing: border-box;
  padding: 0;
}
[type=\\"number\\"]::-webkit-inner-spin-button,
[type=\\"number\\"]::-webkit-outer-spin-button {
  height: auto;
}
[type=\\"search\\"] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
[type=\\"search\\"]::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
details {
  display: block;
}
summary {
  display: list-item;
}
template {
  display: none;
}
[hidden] {
  display: none;
}
body {
  padding-top: 0;
  padding-left: 0;
  padding-bottom: 0;
  padding-right: 0;
  margin-top: 0;
  margin-left: 0;
  margin-bottom: 0;
  margin-right: 0;
  font-family: \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Apple Color Emoji\\",
    \\"Segoe UI Emoji\\", Helvetica, Arial, sans-serif;
  font-size: 0.875rem;
  line-height: 1.4286;
}
* {
  box-sizing: border-box;
}
*:before {
  box-sizing: border-box;
}
*:after {
  box-sizing: border-box;
}
@-webkit-keyframes k1 {
  from {
    color: red;
  }
  to {
    color: blue;
  }
}
@-moz-keyframes k1 {
  from {
    color: red;
  }
  to {
    color: blue;
  }
}
@keyframes k1 {
  from {
    color: red;
  }
  to {
    color: blue;
  }
}
.a {
  background-color: #fff;
}
.b {
  color: #252423;
}
.c {
  text-align: left;
}
.d ::-webkit-scrollbar {
  height: 1rem;
}
.e ::-webkit-scrollbar {
  width: 1rem;
}
.f ::-webkit-scrollbar:disabled {
  display: none;
}
.g ::-webkit-scrollbar-thumb {
  border-top-left-radius: 0.5625rem;
}
.h ::-webkit-scrollbar-thumb {
  border-top-right-radius: 0.5625rem;
}
.i ::-webkit-scrollbar-thumb {
  border-bottom-right-radius: 0.5625rem;
}
.j ::-webkit-scrollbar-thumb {
  border-bottom-left-radius: 0.5625rem;
}
.k ::-webkit-scrollbar-thumb {
  border-top-style: solid;
}
.l ::-webkit-scrollbar-thumb {
  border-right-style: solid;
}
.m ::-webkit-scrollbar-thumb {
  border-bottom-style: solid;
}
.n ::-webkit-scrollbar-thumb {
  border-left-style: solid;
}
.o ::-webkit-scrollbar-thumb {
  border-top-width: 0.25rem;
}
.p ::-webkit-scrollbar-thumb {
  border-right-width: 0.25rem;
}
.q ::-webkit-scrollbar-thumb {
  border-bottom-width: 0.25rem;
}
.r ::-webkit-scrollbar-thumb {
  border-left-width: 0.25rem;
}
.s ::-webkit-scrollbar-thumb {
  border-top-color: transparent;
}
.t ::-webkit-scrollbar-thumb {
  border-right-color: transparent;
}
.u ::-webkit-scrollbar-thumb {
  border-bottom-color: transparent;
}
.v ::-webkit-scrollbar-thumb {
  border-left-color: transparent;
}
.w ::-webkit-scrollbar-thumb {
  background-clip: content-box;
}
.x ::-webkit-scrollbar-thumb {
  background-color: #c8c6c4;
}
.y ::-webkit-scrollbar-thumb:hover {
  background-color: #979593;
}
.z ::-webkit-scrollbar-thumb:hover {
  border-top-style: solid;
}
.ab ::-webkit-scrollbar-thumb:hover {
  border-right-style: solid;
}
.ac ::-webkit-scrollbar-thumb:hover {
  border-bottom-style: solid;
}
.ae ::-webkit-scrollbar-thumb:hover {
  border-left-style: solid;
}
.af ::-webkit-scrollbar-thumb:hover {
  border-top-width: 0.125rem;
}
.ag ::-webkit-scrollbar-thumb:hover {
  border-right-width: 0.125rem;
}
.ah ::-webkit-scrollbar-thumb:hover {
  border-bottom-width: 0.125rem;
}
.ai ::-webkit-scrollbar-thumb:hover {
  border-left-width: 0.125rem;
}
.aj ::-webkit-scrollbar-thumb:hover {
  border-top-color: transparent;
}
.ak ::-webkit-scrollbar-thumb:hover {
  border-right-color: transparent;
}
.al ::-webkit-scrollbar-thumb:hover {
  border-bottom-color: transparent;
}
.am ::-webkit-scrollbar-thumb:hover {
  border-left-color: transparent;
}
.an ::-webkit-scrollbar-track {
  background-color: transparent;
}
.ao {
  display: inline-block;
}
.ap {
  animation-name: k1;
}
.aq {
  animation-duration: 5s;
}
.ar {
  speak: none;
}
.as {
  vertical-align: middle;
}
.at {
  padding-top: 0.25rem;
}
.au {
  padding-right: 0.25rem;
}
.av {
  padding-bottom: 0.25rem;
}
.aw {
  padding-left: 0.25rem;
}
.ax {
  box-shadow: 0 0 0 0.05rem currentColor inset;
}
.ay {
  border-top-left-radius: 50%;
}
.az {
  border-top-right-radius: 50%;
}
.ba {
  border-bottom-right-radius: 50%;
}
.bb {
  border-bottom-left-radius: 50%;
}
.bc {
  align-items: center;
  -webkit-box-align: center;
}
.bd {
  box-sizing: content-box;
}
.be {
  justify-content: center;
  -webkit-box-pack: center;
}
.bf {
  font-size: 1rem;
}
.bg {
  line-height: 1;
}
.bh {
  width: 1rem;
}
.bi {
  height: 1rem;
}
.bj::before {
  content: ?;
}
.bk {
  transform: rotate(0deg);
}
.bl {
  display: none;
}
.bm {
  display: block;
}
.bn {
  fill: currentColor;
}


<div
  className=ui-provider__box a b c d e f g h i j k l m n o p q r s t u v w x y z ab ac ae af ag ah ai aj ak al am an
  dir=ltr
>
  <div className=ui-provider__box a b c d e f g h i j k l m n o p q r s t u v w x y z ab ac ae af ag ah ai aj ak al am an>
    <div className=ui-animation ao>
      <span
        className=ui-icon ar as at au av aw ax ay az ba bb bc bd ao be bf bg bh bi bj bk ap aq
        role=img
        aria-hidden=true
      />
    </div>
  </div>
</div>;
"
`;

exports[`keyframe returning css fallback value 1`] = `
"@font-face {
  font-weight: 400;
  src: url(\\"https://static2.sharepointonline.com/files/fabric/assets/fonts/segoeui-westeuropean/segoeui-regular.woff2\\")
    format(\\"woff2\\");
  font-family: \\"Segoe UI\\";
}
@font-face {
  font-weight: 600;
  src: url(\\"https://static2.sharepointonline.com/files/fabric/assets/fonts/segoeui-westeuropean/segoeui-semibold.woff2\\")
    format(\\"woff2\\");
  font-family: \\"Segoe UI\\";
}
@font-face {
  font-weight: 700;
  src: url(\\"https://static2.sharepointonline.com/files/fabric/assets/fonts/segoeui-westeuropean/segoeui-bold.woff2\\")
    format(\\"woff2\\");
  font-family: \\"Segoe UI\\";
} /*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */
html {
  line-height: 1.15;
  -webkit-text-size-adjust: 100%;
}
body {
  margin: 0;
}
main {
  display: block;
}
h1 {
  font-size: 2em;
  margin: 0.67em 0;
}
hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}
pre {
  font-family: monospace, monospace;
  font-size: 1em;
}
a {
  background-color: transparent;
}
abbr[title] {
  border-bottom: none;
  text-decoration: underline;
  text-decoration: underline dotted;
}
b,
strong {
  font-weight: bolder;
}
code,
kbd,
samp {
  font-family: monospace, monospace;
  font-size: 1em;
}
small {
  font-size: 80%;
}
sub,
sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
sub {
  bottom: -0.25em;
}
sup {
  top: -0.5em;
}
img {
  border-style: none;
}
button,
input,
optgroup,
select,
textarea {
  font-family: inherit;
  font-size: 100%;
  line-height: 1.15;
  margin: 0;
}
button,
input {
  overflow: visible;
}
button,
select {
  text-transform: none;
}
[type=\\"button\\"],
[type=\\"reset\\"],
[type=\\"submit\\"],
button {
  -webkit-appearance: button;
}
[type=\\"button\\"]::-moz-focus-inner,
[type=\\"reset\\"]::-moz-focus-inner,
[type=\\"submit\\"]::-moz-focus-inner,
button::-moz-focus-inner {
  border-style: none;
  padding: 0;
}
[type=\\"button\\"]:-moz-focusring,
[type=\\"reset\\"]:-moz-focusring,
[type=\\"submit\\"]:-moz-focusring,
button:-moz-focusring {
  outline: 1px dotted ButtonText;
}
fieldset {
  padding: 0.35em 0.75em 0.625em;
}
legend {
  box-sizing: border-box;
  color: inherit;
  display: table;
  max-width: 100%;
  padding: 0;
  white-space: normal;
}
progress {
  vertical-align: baseline;
}
textarea {
  overflow: auto;
}
[type=\\"checkbox\\"],
[type=\\"radio\\"] {
  box-sizing: border-box;
  padding: 0;
}
[type=\\"number\\"]::-webkit-inner-spin-button,
[type=\\"number\\"]::-webkit-outer-spin-button {
  height: auto;
}
[type=\\"search\\"] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
[type=\\"search\\"]::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
details {
  display: block;
}
summary {
  display: list-item;
}
template {
  display: none;
}
[hidden] {
  display: none;
}
body {
  padding-top: 0;
  padding-left: 0;
  padding-bottom: 0;
  padding-right: 0;
  margin-top: 0;
  margin-left: 0;
  margin-bottom: 0;
  margin-right: 0;
  font-family: \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Apple Color Emoji\\",
    \\"Segoe UI Emoji\\", Helvetica, Arial, sans-serif;
  font-size: 0.875rem;
  line-height: 1.4286;
}
* {
  box-sizing: border-box;
}
*:before {
  box-sizing: border-box;
}
*:after {
  box-sizing: border-box;
}
@-webkit-keyframes k1 {
  0% {
    opacity: 0.28;
  }
  100% {
    opacity: 0.28;
  }
}
@-moz-keyframes k1 {
  0% {
    opacity: 0.28;
  }
  100% {
    opacity: 0.28;
  }
}
@keyframes k1 {
  0% {
    opacity: 0.28;
  }
  100% {
    opacity: 0.28;
  }
}
.a {
  background-color: #fff;
}
.b {
  color: #252423;
}
.c {
  text-align: left;
}
.d ::-webkit-scrollbar {
  height: 1rem;
}
.e ::-webkit-scrollbar {
  width: 1rem;
}
.f ::-webkit-scrollbar:disabled {
  display: none;
}
.g ::-webkit-scrollbar-thumb {
  border-top-left-radius: 0.5625rem;
}
.h ::-webkit-scrollbar-thumb {
  border-top-right-radius: 0.5625rem;
}
.i ::-webkit-scrollbar-thumb {
  border-bottom-right-radius: 0.5625rem;
}
.j ::-webkit-scrollbar-thumb {
  border-bottom-left-radius: 0.5625rem;
}
.k ::-webkit-scrollbar-thumb {
  border-top-style: solid;
}
.l ::-webkit-scrollbar-thumb {
  border-right-style: solid;
}
.m ::-webkit-scrollbar-thumb {
  border-bottom-style: solid;
}
.n ::-webkit-scrollbar-thumb {
  border-left-style: solid;
}
.o ::-webkit-scrollbar-thumb {
  border-top-width: 0.25rem;
}
.p ::-webkit-scrollbar-thumb {
  border-right-width: 0.25rem;
}
.q ::-webkit-scrollbar-thumb {
  border-bottom-width: 0.25rem;
}
.r ::-webkit-scrollbar-thumb {
  border-left-width: 0.25rem;
}
.s ::-webkit-scrollbar-thumb {
  border-top-color: transparent;
}
.t ::-webkit-scrollbar-thumb {
  border-right-color: transparent;
}
.u ::-webkit-scrollbar-thumb {
  border-bottom-color: transparent;
}
.v ::-webkit-scrollbar-thumb {
  border-left-color: transparent;
}
.w ::-webkit-scrollbar-thumb {
  background-clip: content-box;
}
.x ::-webkit-scrollbar-thumb {
  background-color: #c8c6c4;
}
.y ::-webkit-scrollbar-thumb:hover {
  background-color: #979593;
}
.z ::-webkit-scrollbar-thumb:hover {
  border-top-style: solid;
}
.ab ::-webkit-scrollbar-thumb:hover {
  border-right-style: solid;
}
.ac ::-webkit-scrollbar-thumb:hover {
  border-bottom-style: solid;
}
.ae ::-webkit-scrollbar-thumb:hover {
  border-left-style: solid;
}
.af ::-webkit-scrollbar-thumb:hover {
  border-top-width: 0.125rem;
}
.ag ::-webkit-scrollbar-thumb:hover {
  border-right-width: 0.125rem;
}
.ah ::-webkit-scrollbar-thumb:hover {
  border-bottom-width: 0.125rem;
}
.ai ::-webkit-scrollbar-thumb:hover {
  border-left-width: 0.125rem;
}
.aj ::-webkit-scrollbar-thumb:hover {
  border-top-color: transparent;
}
.ak ::-webkit-scrollbar-thumb:hover {
  border-right-color: transparent;
}
.al ::-webkit-scrollbar-thumb:hover {
  border-bottom-color: transparent;
}
.am ::-webkit-scrollbar-thumb:hover {
  border-left-color: transparent;
}
.an ::-webkit-scrollbar-track {
  background-color: transparent;
}
.ao {
  display: inline-block;
}
.ap {
  animation-name: k1;
}
.aq {
  speak: none;
}
.ar {
  vertical-align: middle;
}
.as {
  padding-top: 0.25rem;
}
.at {
  padding-right: 0.25rem;
}
.au {
  padding-bottom: 0.25rem;
}
.av {
  padding-left: 0.25rem;
}
.aw {
  box-shadow: 0 0 0 0.05rem currentColor inset;
}
.ax {
  border-top-left-radius: 50%;
}
.ay {
  border-top-right-radius: 50%;
}
.az {
  border-bottom-right-radius: 50%;
}
.ba {
  border-bottom-left-radius: 50%;
}
.bb {
  align-items: center;
  -webkit-box-align: center;
}
.bc {
  box-sizing: content-box;
}
.bd {
  justify-content: center;
  -webkit-box-pack: center;
}
.be {
  font-size: 1rem;
}
.bf {
  line-height: 1;
}
.bg {
  width: 1rem;
}
.bh {
  height: 1rem;
}
.bi::before {
  content: ?;
}
.bj {
  transform: rotate(0deg);
}
.bk {
  display: none;
}
.bl {
  display: block;
}
.bm {
  fill: currentColor;
}


<div
  className=ui-provider__box a b c d e f g h i j k l m n o p q r s t u v w x y z ab ac ae af ag ah ai aj ak al am an
  dir=ltr
>
  <div className=ui-provider__box a b c d e f g h i j k l m n o p q r s t u v w x y z ab ac ae af ag ah ai aj ak al am an>
    <div className=ui-animation ao>
      <span
        className=ui-icon aq ar as at au av aw ax ay az ba bb bc ao bd be bf bg bh bi bj ap
        role=img
        aria-hidden=true
      />
    </div>
  </div>
</div>;
"
`;

exports[`rtl 1`] = `
"

<div className=ui-provider__box dir=rtl>
  <span className=ui-text dir=auto>
    Hello
  </span>
</div>;
"
`;
